src/Components/WebView/Platforms/Wpf/src/BlazorWebView.cs

```diff
@@ -0,0 +1,150 @@
+
+        private void OnServicesPropertyChanged(DependencyPropertyChangedEventArgs e) => StartWebViewCoreIfPossible();
+
+        private static void OnHostPagePropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e) => ((BlazorWebView)d).OnHostPagePropertyChanged(e);
+
+        private void OnHostPagePropertyChanged(DependencyPropertyChangedEventArgs e) => StartWebViewCoreIfPossible();
+
+        private bool RequiredStartupPropertiesSet =>
+            _webview != null &&
+            HostPage != null &&
+            Services != null;
+
+        public override void OnApplyTemplate()
+        {
+            // Called when the control is created after its child control (the WebView2) is created from the Template property
+            base.OnApplyTemplate();
+
+            if (_webview == null)
+            {
+                _webview = (WebView2Control)GetTemplateChild(webViewTemplateChildName);
+                StartWebViewCoreIfPossible();
+            }
+        }
+
+        protected override void OnInitialized(EventArgs e)
+        {
+            // Called when BeginInit/EndInit are used, such as when creating the control from XAML
+            base.OnInitialized(e);
+            StartWebViewCoreIfPossible();
+        }
+
+        private void StartWebViewCoreIfPossible()
+        {
+            if (!RequiredStartupPropertiesSet || _webviewManager != null)
+            {
+                return;
+            }
+
+            // We assume the host page is always in the root of the content directory, because it's
+            // unclear there's any other use case. We can add more options later if so.
+            var contentRootDir = Path.GetDirectoryName(Path.GetFullPath(HostPage));
+            var fileProvider = new PhysicalFileProvider(contentRootDir);
+
+            _webviewManager = new WebView2WebViewManager(new WpfWeb2ViewWrapper(_webview), Services, WpfDispatcher.Instance, fileProvider);
+            foreach (var rootComponent in RootComponents)
+            {
+                // Since the page isn't loaded yet, this will always complete synchronously
+                _ = rootComponent.AddToWebViewManagerAsync(_webviewManager);
+            }
+            _webviewManager.Navigate("/");
+        }
+
+        private void HandleRootComponentsCollectionChanged(object sender, NotifyCollectionChangedEventArgs eventArgs)
+        {
+            // If we haven't initialized yet, this is a no-op
+            if (_webviewManager != null)
+            {
+                // Dispatch because this is going to be async, and we want to catch any errors
+                WpfDispatcher.Instance.InvokeAsync(async () =>
+                {
+                    var newItems = eventArgs.OldItems.Cast<RootComponent>();
+                    var oldItems = eventArgs.NewItems.Cast<RootComponent>();
```

Comment:
Shouln't this be like this?

```suggestion
                    var newItems = eventArgs.NewItems.Cast<RootComponent>();
                    var oldItems = eventArgs.OldItems.Cast<RootComponent>();
```

Also both `foreach` loops are doing `newItems.Except(oldItems)` shloudn't one be the other way around?

Reply:
Well spotted :) Fixed.